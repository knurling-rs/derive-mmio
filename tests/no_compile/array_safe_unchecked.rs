#[derive(derive_mmio::Mmio)]
#[repr(C)]
struct Uart {
    array: [u32; 2],
}

fn main() {
    let mut uart = Uart { array: [0; 2] };

    // Safety: We're pointing at a real object
    let mmio_uart = unsafe { Uart::new_mmio(core::ptr::addr_of_mut!(uart)) };

    // we cannot safely access the array field without bounds checks

    let _inner_bank = mmio_uart.read_array_unchecked(5);
}
