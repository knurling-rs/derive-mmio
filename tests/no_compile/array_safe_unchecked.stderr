error[E0133]: call to unsafe function `MmioUart::<'_>::read_array_unchecked` is unsafe and requires unsafe function or block
  --> tests/no_compile/array_safe_unchecked.rs:15:23
   |
15 |     let _inner_bank = mmio_uart.read_array_unchecked(5);
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ call to unsafe function
   |
   = note: consult the function's documentation for information on how to avoid undefined behavior
